### PR TITLE
RegCM: intel FC 'hint', 4.7.0 intel warning, fcflags, fix extensions.

### PR DESCRIPTION
**Pull request by bebosudo**
'any_combination_of' expects a sequence of objects, but was getting a single
iterable object (PR #9481).

This should fix and improve #9934, which was messed up by a fetch from `develop`.

There are some new versions of RegCM available:
```python
    version('4.7.0-SVN6916',
            sha256='cad66b760464b513afd318eb28a6820a0d4c8c55d3b23a34d06cb9c1b57aa0bc',
            url='http://clima-dods.ictp.it/Users/ggiulian/RegCM-SVN6916.tar.gz')
    version('4.7.0-SVN6884',
            sha256='fd4eb0fbe911c6c4dacbaaf2a952a45ac8c1fa54a5d905972ffe592e47818c00',
            url='http://clima-dods.ictp.it/Users/ggiulian/RegCM-SVN6884.tar.gz')
```
A new "official" version has not been released yet.

The latest version provides a partial support to `pnetcdf` (direct support, without using `netcdf`, would cost more developing time).

### CODE REVIEW

**Comment by bebosudo** [comment]
@alalazo I've unpacked

**Comment by alalazo** [comment]
@bebosudo Thanks for the fix! That was obviously a typo I introduced in #9481

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 
 
 class Regcm(AutotoolsPackage):
-    """RegCM ICTP Regional Climate Model."""
+    '''RegCM, ICTP Regional Climate Model (https://ictp.it).'''
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/regcm/package.py]: Apologies for the OCD, but every other package uses double quotes. Can you maintain them?

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 2]
```diff
@@ -23,17 +23,24 @@ class Regcm(AutotoolsPackage):
 
     # On Intel and PGI compilers, multiple archs can be built at the same time,
     # producing a so-called fat binary. Unfortunately, gcc builds only the last
-    # architecture provided (in the configure), so we allow a single arch.
+    # architecture provided (in the configure), so we allow a single arch on
+    # GCC. Only GCC and Intel are supported.
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/regcm/package.py]: I don't really understand the comment. At the beginning it seems PGI compilers are allowed. At the end it seems that only GCC and Intel can compile this package.

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 3]
```diff
@@ -56,20 +63,54 @@ def flag_handler(self, name, flags):
         return (None, None, flags)
 
     def configure_args(self):
+        if self.spec.satisfies(r'@4.7.0 %pgi'):
+            raise InstallError('Intel compiler not working with this specific '
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/regcm/package.py]: Probably you meant PGI compiler instead of Intel compiler? On top of that you can do checks like this one at the end of the concretization phase by using a `conflict` directive:
```python
conflict('%pgi', when='@4.7.0', msg=error_message)
```

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 4]
```diff
@@ -56,20 +63,54 @@ def flag_handler(self, name, flags):
...
+                # supported by RegCM 4.7.x.
+                raise InstallError('Architecture optimizations are available '
+                                   'only for GCC and Intel compilers.')
+
+            if len(optimizations) > 1 and self.spec.satisfies(r'%gcc'):
+                # https://github.com/spack/spack/issues/974
+                raise InstallError('The GCC compiler does not support '
+                                   'multiple architecture optimizations.')
+
+            args += ('--enable-' + ext for ext in optimizations)
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/regcm/package.py]: Given that the package follows a very structured naming convention, it should work seamlessly if you say:
```python
args += self.enable_or_disable('extension')
```

File: var/spack/repos/builtin/packages/regcm/package.py [hunk 5]
```diff
@@ -56,20 +63,54 @@ def flag_handler(self, name, flags):
...
+
         for opt in ('debug', 'profile', 'singleprecision'):
-            if '+{0}'.format(opt) in self.spec:
+            if ('+' + opt) in self.spec:
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/regcm/package.py]: Similarly here you can use:
```python
args += self.enable_or_disable(opt)
```

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 1 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: I'll fix it in a new commit.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 2 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: See this discussion with @adamjstewart: https://github.com/spack/spack/pull/9934#pullrequestreview-178034960
and also this #974 (in particular this comment: https://github.com/spack/spack/issues/974#issuecomment-225247436)

Basically, Intel and PGI can build multi-arch binaries, while GCC cannot.
Moreover, this package (RegCM) has optimizations defined only for Intel and GCC, but should be buildable also with other compilers.

To sum up:
- with intel, a user should be able to build a single executables for all the combinations of architectures they want (using `--extension=knl,skl,bdw,nhl`);
- with gcc, a user is allowed to build an executable using a single optimization;
- with any other compiler, no _extensions/optimizations_ are supported.

That's also the reason for those "strange" checks in `configure_args`.

I can rephrase that comment better.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 3 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Yes, you're right, it's a typo.
Thanks for the tip on `msg`, I'll use it.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 4 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Does this work with multiple values?
I mean: if the user gives the following spec: `extension=knl,bdw` (which works only on Intel, see discussion above), does `enable_or_disable` produce the following:
`--enable-knl --enable-bdw --disable-nhl --disable-skl`
?

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 5 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Great!

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 4 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Yes, it works also with multiple values.

File: var/spack/repos/builtin/packages/regcm/package.py (hunk 2 above)
Comment (bebosudo) [reply var/spack/repos/builtin/packages/regcm/package.py]: Hope the comment in commit 75fdc14f clarifies.

**Comment by bebosudo** [comment]
Great, thanks!

**Comment by bebosudo** [comment]
I've just discovered that the configure of this package doesn't comply with its own rules.
If you pass `--disable-FEATURE`, it will ignore your disable, and consider it like `--enable-FEATURE`, so `enable_or_disable` cannot be used, unfortunately :disappointed:

@alalazo do you think I should "reopen" this issue? (don't even know whether is it possible on github)
or should I just open a new one?

I'd also like to include a fix for #10163.

**Comment by adamjstewart** [comment]
Just create a new PR.

**Comment by bebosudo** [comment]
@adamjstewart done! Opened #10863, but I have a problem with a type 'run' of dependency, could you have a look at it?
